hubcast/utils/git.py

```diff
@@ -1,17 +1,31 @@
-import os
 import subprocess
+from dataclasses import dataclass, field
+from typing import Dict, List
 
-# get configuration from environment.
-GIT_REPO_PATH = os.environ.get("HC_GIT_REPO_PATH")
 
+# https://git-scm.com/docs/gitcredentials#_custom_helpers
+def credential_helper(remote: str, url: str) -> str:
+    remote = remote.upper()
 
-def git(args):
-    """Executes a git command on the host system."""
-    result = subprocess.run(
-        ["git", "-C", f"{GIT_REPO_PATH}"] + args.split(),
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-        text=True,
-        check=True,
-    )
-    return result
+    username_str = f"username=${remote}_USERNAME"
+    pass_str = f"password=${remote}_TOKEN"
+
+    helper_func = f"""!f() {{ echo "{username_str}"; echo "{pass_str}"; }};f"""
+    return f"""credential.{url}.helper={helper_func}"""
+
+
+@dataclass
+class Git:
+    env: Dict[str, str] = field(default_factory=lambda: {})
+    flags: List[str] = field(default_factory=lambda: [])
+
+    def __call__(self, args: List[str]):
```

Comment:
Not critical, but I liked calling `git` with a string like:

```python
git('pull github foo')
```

I thought it looked nice and hid the detail that `subprocess` was handling the request under the hood.

If we wanted to support that still we could use stdlib [`shlex.split`](https://docs.python.org/3/library/shlex.html#shlex.split) and modify the subprocess call:

```python
    def __call__(self, cmd: str):
        return subprocess.run(
            ["git", *self.flags] + shlex.split(cmd),
            ...
        )
```

Again, merely personal preference so not critical to change :)

Reply:
Oh for sure, totally willing to going back to a string. I'd kinda like to merge this first, but I'll create an issue for a follow up PR to move back to a string based approach.